In [110]:
import logging

logging.getLogger('fastf1').setLevel(logging.ERROR)

import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)

In [111]:
import fastf1
import pandas as pd
import numpy as np
from pathlib import Path
import os

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

fastf1.Cache.enable_cache('cache')

In [112]:
df_drivers = pd.read_csv('data/Model2/drivers.csv').convert_dtypes()
df_laps = pd.read_csv('data/Model2/laps.csv').convert_dtypes()
df_track_status = pd.read_csv('data/Model2/track_status.csv').convert_dtypes()
df_weather = pd.read_csv('data/Model2/weather.csv').convert_dtypes()

C:\Users\ognjeni\AppData\Local\Temp\ipykernel_20860\2993207287.py:2: DtypeWarning: Columns (18) have mixed types. Specify dtype option on import or set low_memory=False.
  df_laps = pd.read_csv('data/Model2/laps.csv').convert_dtypes()


In [113]:
for col in df_laps.select_dtypes(include=['string','boolean']).columns:
    if df_laps[col].dtype.name == 'boolean':
        df_laps[col] = df_laps[col].astype('float')
    else:
        df_laps[col] = df_laps[col].astype('object')

In [114]:
df_laps = df_laps.replace({pd.NA: np.nan})

In [57]:
df_laps.head()

,Time,Driver,DriverNumber,LapTime,LapNumber,Stint,PitOutTime,PitInTime,Sector1Time,Sector2Time,Sector3Time,Sector1SessionTime,Sector2SessionTime,Sector3SessionTime,SpeedI1,SpeedI2,SpeedFL,SpeedST,IsPersonalBest,Compound,TyreLife,FreshTyre,Team,LapStartTime,LapStartDate,TrackStatus,Position,Deleted,DeletedReason,FastF1Generated,IsAccurate,Season,Round,EventName
0,0 days 00:34:06.302000,GAS,10,0 days 00:01:19.106000,1,1,NaN,NaN,NaN,0 days 00:00:33.352000,0 days 00:00:23.247000,NaN,0 days 00:33:43.079000,0 days 00:34:06.511000,309,217,274,290,0.0,MEDIUM,1,1.0,AlphaTauri,0 days 00:32:47.006000,<NA>,1,12,<NA>,<NA>,0.0,0.0,2020,0,Pre-Season Test 1
1,0 days 00:35:18.714000,GAS,10,0 days 00:01:12.412000,2,1,NaN,NaN,0 days 00:00:17.960000,0 days 00:00:32.228000,0 days 00:00:22.224000,0 days 00:34:24.262000,0 days 00:34:56.490000,0 days 00:35:18.714000,282,226,270,292,1.0,MEDIUM,2,1.0,AlphaTauri,0 days 00:34:06.302000,<NA>,1,12,<NA>,<NA>,0.0,1.0,2020,0,Pre-Season Test 1
2,0 days 00:36:30.025000,GAS,10,0 days 00:01:11.311000,3,1,NaN,NaN,0 days 00:00:17.513000,0 days 00:00:31.717000,0 days 00:00:22.081000,0 days 00:35:36.227000,0 days 00:36:07.944000,0 days 00:36:30.025000,307,224,277,311,1.0,MEDIUM,3,1.0,AlphaTauri,0 days 00:35:18.714000,<NA>,1,12,<NA>,<NA>,0.0,1.0,2020,0,Pre-Season Test 1
3,0 days 00:37:40.750000,GAS,10,0 days 00:01:10.725000,4,1,NaN,NaN,0 days 00:00:17.410000,0 days 00:00:31.510000,0 days 00:00:21.805000,0 days 00:36:47.435000,0 days 00:37:18.945000,0 days 00:37:40.750000,309,218,275,309,1.0,MEDIUM,4,1.0,AlphaTauri,0 days 00:36:30.025000,<NA>,1,12,<NA>,<NA>,0.0,1.0,2020,0,Pre-Season Test 1
4,0 days 00:38:52.261000,GAS,10,0 days 00:01:11.511000,5,1,NaN,NaN,0 days 00:00:17.464000,0 days 00:00:31.827000,0 days 00:00:22.220000,0 days 00:37:58.214000,0 days 00:38:30.041000,0 days 00:38:52.261000,303,219,268,296,0.0,MEDIUM,5,1.0,AlphaTauri,0 days 00:37:40.750000,<NA>,1,12,<NA>,<NA>,0.0,1.0,2020,0,Pre-Season Test 1


In [7]:
df_drivers.tail()

,DriverNumber,BroadcastName,Abbreviation,DriverId,TeamName,TeamColor,TeamId,FirstName,LastName,FullName,HeadshotUrl,CountryCode,Position,ClassifiedPosition,GridPosition,Q1,Q2,Q3,Time,Status,Points,Laps,Season,Round
2114,77,V BOTTAS,BOT,bottas,Kick Sauber,52e252,sauber,Valtteri,Bottas,Valtteri Bottas,https://media.formula1.com/d_driver_fallback_i...,FIN,16,16,19,<NA>,<NA>,<NA>,0 days 00:00:57.829000,Lapped,0.0,61,2024,18
2115,10,P GASLY,GAS,gasly,Alpine,0093cc,alpine,Pierre,Gasly,Pierre Gasly,https://media.formula1.com/d_driver_fallback_i...,FRA,17,17,18,<NA>,<NA>,<NA>,0 days 00:00:59.059000,Lapped,0.0,61,2024,18
2116,3,D RICCIARDO,RIC,ricciardo,RB,6692FF,rb,Daniel,Ricciardo,Daniel Ricciardo,https://media.formula1.com/d_driver_fallback_i...,AUS,18,18,16,<NA>,<NA>,<NA>,0 days 00:01:29.796000,Lapped,0.0,61,2024,18
2117,20,K MAGNUSSEN,MAG,kevin_magnussen,Haas F1 Team,B6BABD,haas,Kevin,Magnussen,Kevin Magnussen,https://media.formula1.com/d_driver_fallback_i...,DEN,19,19,14,<NA>,<NA>,<NA>,<NA>,Retired,0.0,57,2024,18
2118,23,A ALBON,ALB,albon,Williams,64C4FF,williams,Alexander,Albon,Alexander Albon,https://media.formula1.com/d_driver_fallback_i...,THA,20,R,11,<NA>,<NA>,<NA>,<NA>,Retired,0.0,15,2024,18


In [9]:
df_track_status.head()

,Time,Status,Message,Season,Round
0,0 days 00:07:56.420000,1,AllClear,2020,0
1,0 days 00:56:21.333000,2,Yellow,2020,0
2,0 days 00:56:30.327000,1,AllClear,2020,0
3,0 days 01:01:52.836000,2,Yellow,2020,0
4,0 days 01:02:31.497000,4,SCDeployed,2020,0


In [10]:
df_weather.head()

,Time,AirTemp,Humidity,Pressure,Rainfall,TrackTemp,WindDirection,WindSpeed,Season,Round
0,0 days 00:00:38.669000,27.5,35.1,939.4,False,55.6,99,1.7,2020,0
1,0 days 00:01:38.684000,27.1,36.5,939.4,False,55.1,106,1.9,2020,0
2,0 days 00:02:38.655000,27.0,36.2,939.3,False,54.2,111,2.1,2020,0
3,0 days 00:03:38.669000,26.9,36.1,939.3,False,54.2,165,1.0,2020,0
4,0 days 00:04:38.678000,27.2,35.8,939.3,False,54.2,83,1.2,2020,0


In [12]:
df_laps.dtypes

Time                   object
Driver                 object
DriverNumber            Int64
LapTime                object
LapNumber               Int64
Stint                   Int64
PitOutTime             object
PitInTime              object
Sector1Time            object
Sector2Time            object
Sector3Time            object
Sector1SessionTime     object
Sector2SessionTime     object
Sector3SessionTime     object
SpeedI1                 Int64
SpeedI2                 Int64
SpeedFL                 Int64
SpeedST                 Int64
IsPersonalBest        float64
Compound               object
TyreLife                Int64
FreshTyre             float64
Team                   object
LapStartTime           object
LapStartDate            Int64
TrackStatus             Int64
Position                Int64
Deleted                 Int64
DeletedReason           Int64
FastF1Generated       float64
IsAccurate            float64
Season                  Int64
Round                   Int64
EventName 

Every lap that is longer than 4 minutes will be considered as outlier.

In [115]:
df_laps['LapTime'] = pd.to_timedelta(df_laps['LapTime'])

In [116]:
print('Length of df laps: ', len(df_laps))
print('Amount of laps that have time more than 4 minuts: ', len(df_laps[df_laps['LapTime'] > pd.Timedelta(minutes=4)]))

Length of df laps:  115530
Amount of laps that have time more than 4 minuts:  21


In [117]:
df_laps = df_laps[df_laps['LapTime'] <= pd.Timedelta(minutes=4)]

# Target creation

In [118]:
df_laps['LapTime_ms'] = df_laps['LapTime'].dt.total_seconds() * 1000

In [119]:
df_laps = df_laps.sort_values(['Season', 'Round', 'EventName', 'Driver', 'LapNumber'])

In [120]:
df_laps['target'] = (
    df_laps
    .groupby(['Season', 'Round', 'EventName', 'Driver'])['LapTime_ms']
    .shift(-1)
)

We are dropping na values for target, because, this data represent last laps. Last laps does not have next time lap.

In [125]:
df_laps = df_laps.dropna(subset=['target'])

In [126]:
df = df_laps

In [127]:
df.describe()

,DriverNumber,LapTime,LapNumber,Stint,SpeedI1,SpeedI2,SpeedFL,SpeedST,IsPersonalBest,TyreLife,FreshTyre,LapStartDate,TrackStatus,Position,Deleted,DeletedReason,FastF1Generated,IsAccurate,Season,Round,LapTime_ms,target
count,111310.0,111310,111310.0,111310.0,94779.0,111283.0,107697.0,102705.0,111310.000000,110718.0,111310.000000,0.0,111287.0,111310.0,0.0,0.0,111310.0,111310.000000,111310.0,111310.0,111310.000000,111310.000000
mean,27.6145,0 days 00:01:30.846066624,30.285671,2.103028,255.754197,244.826416,262.023139,293.843902,0.217537,14.488927,0.762959,<NA>,9.952043,9.713081,<NA>,<NA>,0.0,0.873237,2021.994421,9.941353,90846.066625,90564.833447
std,24.504953,0 days 00:00:14.774084974,18.150637,0.986685,48.401857,43.275911,36.488779,30.896939,0.412572,10.063014,0.425269,<NA>,103.086647,5.391221,<NA>,<NA>,0.0,0.332709,1.350289,6.342127,14774.084975,14483.270695
min,1.0,0 days 00:00:55.404000,1.0,1.0,40.0,35.0,1.0,59.0,0.000000,1.0,0.000000,<NA>,1.0,1.0,<NA>,<NA>,0.0,0.000000,2020.0,0.0,55404.000000,55404.000000
25%,10.0,0 days 00:01:20.702250,15.0,1.0,221.0,222.0,245.0,288.0,0.000000,7.0,1.000000,<NA>,1.0,5.0,<NA>,<NA>,0.0,1.000000,2021.0,5.0,80702.250000,80567.250000
50%,20.0,0 days 00:01:29.646500,29.0,2.0,273.0,253.0,272.0,298.0,0.000000,13.0,1.000000,<NA>,1.0,10.0,<NA>,<NA>,0.0,1.000000,2022.0,10.0,89646.500000,89301.500000
75%,44.0,0 days 00:01:39.435000,44.0,3.0,289.0,276.0,286.0,309.0,0.000000,20.0,1.000000,<NA>,1.0,14.0,<NA>,<NA>,0.0,1.000000,2023.0,15.0,99435.000000,99201.000000
max,99.0,0 days 00:03:46.277000,86.0,8.0,359.0,350.0,355.0,362.0,1.000000,77.0,1.000000,<NA>,6712.0,20.0,<NA>,<NA>,0.0,1.000000,2024.0,22.0,226277.000000,149985.000000


# First model

In [128]:
from sklearn.pipeline import Pipeline

from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.impute import SimpleImputer
from sklearn.base import BaseEstimator, TransformerMixin
from sklearn.preprocessing import FunctionTransformer
from sklearn.base import clone
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor
from xgboost import XGBRegressor

In [129]:
from sklearn.metrics import (
    mean_absolute_error,
    root_mean_squared_error,
    r2_score
)

### Some helper functions

In [130]:
class DropColumns(BaseEstimator, TransformerMixin):
    def __init__(self, cols):
        self.cols = cols

    def fit(self, X, y=None):
        return self

    def transform(self, X):
        return X.drop(columns=self.cols, errors='ignore')

In [131]:
def na_to_binary(X):
    return X.notna().astype(int)

In [132]:
class TimeConverter(BaseEstimator, TransformerMixin):
    def __init__(self, cols):
        self.cols = cols

    def fit(self, X, y=None):
        return self

    def transform(self, X):
        X = X.copy()
        for col in self.cols:
            X[col] = pd.to_timedelta(X[col], errors='coerce').dt.total_seconds()
        return X

#### Eval function

In [133]:
# Evaluation func
def eval(df, pipelines, target_col, feature_cols):
    seasons = sorted(df['Season'].unique())
    results = []

    for i in range(1, len(seasons)):
        train_seasons = seasons[:i]
        test_season = seasons[i]

        train_df = df[df['Season'].isin(train_seasons)]
        test_df = df[df['Season'] == test_season]

        X_train = train_df[feature_cols]
        y_train = train_df[target_col]

        X_test = test_df[feature_cols]
        y_test = test_df[target_col]

        print(f"\n=========== ITERATION {i} | TEST SEASON {test_season} ===========")

        models = clone(pipelines)
        for name, model in models.items():
            print(f"\nTraining {name}...")
            model.fit(X_train, y_train)
            y_pred = model.predict(X_test)

            mae = mean_absolute_error(y_test, y_pred)
            rmse = root_mean_squared_error(y_test, y_pred)
            r2 = r2_score(y_test, y_pred)

            results.append({
                'season': test_season,
                'model': name,
                'MAE': mae,
                'RMSE': rmse, 
                'R2': r2
            })
            
            print(f"{name} -> MAE: {mae:.2f}, RMSE: {rmse:.2f}, R2: {r2:.3f}")

    return results

### Data

In [134]:
numerical_cols = ['LapTime', 'LapNumber', 'Stint', 'TyreLife', 'Sector1Time', 'Sector2Time', 'Sector3Time', 'SpeedI1',
                 'SpeedI2', 'SpeedFL', 'SpeedST', 'TrackStatus']
categorical_cols = ['Driver', 'Compound', 'Team', 'FreshTyre', 'EventName']

In [135]:
cols_to_drop = ['Time', 'DriverNumber', 'Sector1SessionTime', 'Sector2SessionTime', 'Sector3SessionTime', 'IsPersonalBest', 
                'LapStartTime', 'LapStartDate', 'Position', 'Deleted', 'DeletedReason', 'FastF1Generated', 'IsAccurate',
                'Season', 'Round', 'NextPosition']

In [136]:
binary_features = ['PitInTime', 'PitOutTime']

In [137]:
preprocessor = Pipeline(steps=[
    ('drop_cols', DropColumns(cols_to_drop)),
    ('convertor', TimeConverter([
        'LapTime', 'Sector1Time', 'Sector2Time', 'Sector3Time'
    ])),
    ('preprocess', ColumnTransformer(
        transformers=[
            
            ('num', Pipeline([
                ('imputer', SimpleImputer(strategy='mean')),
                ('scaler', StandardScaler())
            ]), numerical_cols),
            
            ('cat', Pipeline([
                ('imputer', SimpleImputer(strategy='most_frequent')),
                ('encoder', OneHotEncoder(handle_unknown='ignore'))
            ]), categorical_cols),
            
            ('bin_transform', Pipeline([
                ('to_binary', FunctionTransformer(na_to_binary))
            ]), binary_features)
        ]
    ))
])

In [138]:
pipeline_lr_I = Pipeline(steps=[
    ("preprocessing", preprocessor),
    ("model", LinearRegression())
])

pipeline_rf_I = Pipeline(steps=[
    ("preprocessing", preprocessor),
    ("model", RandomForestRegressor(
        n_estimators=100,
        random_state=42,
        n_jobs=-1
    ))
])

pipeline_xgb_I = Pipeline(steps=[
    ("preprocessing", preprocessor),
    ("model", XGBRegressor(
        n_estimators=300,
        learning_rate=0.05,
        max_depth=6,
        subsample=0.8,
        colsample_bytree=0.8,
        random_state=42
    ))
])

In [139]:
pipelines = {
    "Linear Regression": pipeline_lr_I,
    "Random Forest": pipeline_rf_I,
    "XGBoost": pipeline_xgb_I
}

In [140]:
target_col = 'target'
feature_cols = numerical_cols + categorical_cols + binary_features

In [142]:
results = eval(df, pipelines, target_col, feature_cols)

=========== ITERATION 1 | TEST SEASON 2021 ===========<br/>

Linear Regression -> MAE: 5323.40, RMSE: 8418.83, R2: 0.626<br/>
Random Forest -> MAE: 4886.03, RMSE: 8211.56, R2: 0.644<br/>
XGBoost -> MAE: 4161.69, RMSE: 6955.54, R2: 0.745<br/><br/>

=========== ITERATION 2 | TEST SEASON 2022 ===========<br/>

Linear Regression -> MAE: 3600.96, RMSE: 6968.64, R2: 0.787<br/>
Random Forest -> MAE: 3608.07, RMSE: 7092.53, R2: 0.779<br/>
XGBoost -> MAE: 3329.28, RMSE: 6553.11, R2: 0.811<br/><br/>

=========== ITERATION 3 | TEST SEASON 2023 ===========<br/>

Linear Regression -> MAE: 3407.77, RMSE: 6453.92, R2: 0.770<br/>
Random Forest -> MAE: 2356.38, RMSE: 5349.76, R2: 0.842<br/>
XGBoost -> MAE: 2133.46, RMSE: 5018.17, R2: 0.861<br/><br/>

=========== ITERATION 4 | TEST SEASON 2024 ===========<br/>

Linear Regression -> MAE: 2424.15, RMSE: 5101.60, R2: 0.831<br/>
Random Forest -> MAE: 1911.16, RMSE: 4427.53, R2: 0.873<br/>
XGBoost -> MAE: 1610.12, RMSE: 3924.78, R2: 0.900<br/>


# Second model

### Adding some new features

In [210]:
# Class that has method that generate new feature - previous lap time.

class PrevLap(BaseEstimator, TransformerMixin):
    def __init__(
        self, 
        season_col='Season', 
        round_col='Round',
        driver_col='Driver',
        laptimems_col='LapTime_ms'
    ):
        self.season_col = season_col
        self.round_col = round_col
        self.driver_col = driver_col
        self.laptimems_col = laptimems_col

    def fit(self, X, y=None):
        return self

    def transform(self, X):
        df = X.copy()
        df['prev_lap_ms'] = df.groupby([self.season_col, self.round_col, self.driver_col])[self.laptimems_col].shift(1)
        return df

In [201]:
# Class that has method that creates new feature which represents difference between prevoius lap and current lap time in miliseconds.

class LapDelta(BaseEstimator, TransformerMixin):
    def __init__(
        self,
        laptimems_col='LapTime_ms',
        prevlapms_col='prev_lap_ms'
    ):
        self.laptimems_col=laptimems_col
        self.prevlapms_col=prevlapms_col

    def fit(self, X, y=None):
        return self

    def transform(self, X):
        df = X.copy()
        df['lap_diff_ms'] = df[self.laptimems_col] - df[self.prevlapms_col]
        return df

In [255]:
# Class that has method that generate feature for pace - last 3 laps.

class RollingPace(BaseEstimator, TransformerMixin):
    def __init__(
        self,
        season_col='Season', 
        round_col='Round',
        driver_col='Driver',
        laptimems_col='LapTime_ms'
    ):
        self.season_col = season_col
        self.round_col = round_col
        self.driver_col = driver_col
        self.laptimems_col = laptimems_col

    def fit(self, X, y=None):
        return self

    def transform(self, X):
        df = X.copy()
        df['pace_last_3'] = (
            df.groupby([self.season_col, self.round_col, self.driver_col])[self.laptimems_col]
            .shift(1)
            .rolling(3)
            .mean()
        )

        return df

In [232]:
df.dtypes

Time                           object
Driver                         object
DriverNumber                    Int64
LapTime               timedelta64[ns]
LapNumber                       Int64
Stint                           Int64
PitOutTime                     object
PitInTime                      object
Sector1Time                    object
Sector2Time                    object
Sector3Time                    object
Sector1SessionTime             object
Sector2SessionTime             object
Sector3SessionTime             object
SpeedI1                         Int64
SpeedI2                         Int64
SpeedFL                         Int64
SpeedST                         Int64
IsPersonalBest                float64
Compound                       object
TyreLife                        Int64
FreshTyre                     float64
Team                           object
LapStartTime                   object
LapStartDate                    Int64
TrackStatus                     Int64
Position    

### Preparing pipelines

In [256]:
numerical_cols = ['LapTime', 'LapNumber', 'Stint', 'TyreLife', 'Sector1Time', 'Sector2Time', 'Sector3Time', 'SpeedI1',
                 'SpeedI2', 'SpeedFL', 'SpeedST', 'TrackStatus', 'Season', 'Round', 'LapTime_ms']

categorical_cols = ['Driver', 'Compound', 'Team', 'FreshTyre', 'EventName']

binary_features = ['PitInTime', 'PitOutTime']

cols_to_drop = ['Time', 'DriverNumber', 'Sector1SessionTime', 'Sector2SessionTime', 'Sector3SessionTime', 'IsPersonalBest', 
                'LapStartTime', 'LapStartDate', 'Position', 'Deleted', 'DeletedReason', 'FastF1Generated', 'IsAccurate',
                'NextPosition']

# after engineering
num_cols_after_engineering = numerical_cols + ['prev_laps_ms', 'lap_diff_ms', 'pace_last_3']

In [257]:
feature_engineering_II = Pipeline(steps=[
    ('convertor', TimeConverter([
        'LapTime', 'Sector1Time', 'Sector2Time', 'Sector3Time'
    ])),
    ('prev_lap', PrevLap()),
    ('lap_delta', LapDelta()),
    ('rolling_pace', RollingPace()),
    ('drop_cols',  DropColumns(cols_to_drop))
])

column_preprocess_II = ColumnTransformer(
    transformers=[
        ('num', Pipeline([
            ('imputer', SimpleImputer(strategy='mean')),
            ('scaler', StandardScaler())
        ]), num_cols_after_engineering),

        ('cat', Pipeline([
            ('imputer', SimpleImputer(strategy='most_frequent')),
            ('encoder', OneHotEncoder(handle_unknown='ignore'))
        ]), categorical_cols)
    ],
    remainder='drop'
)

In [258]:
preprocessor_II = Pipeline(steps=[
    ('engineering', feature_engineering_II),
    ('transformers', column_preprocess_II)
])

In [259]:
pipeline_lr_II = Pipeline(steps=[
    ("preprocessing", preprocessor_II),
    ("model", LinearRegression())
])

pipeline_rf_II = Pipeline(steps=[
    ("preprocessing", preprocessor_II),
    ("model", RandomForestRegressor(
        n_estimators=100,
        random_state=42,
        n_jobs=-1
    ))
])

pipeline_xgb_II = Pipeline(steps=[
    ("preprocessing", preprocessor_II),
    ("model", XGBRegressor(
        n_estimators=300,
        learning_rate=0.05,
        max_depth=6,
        subsample=0.8,
        colsample_bytree=0.8,
        random_state=42
    ))
])

In [260]:
pipelines_II = {
    "Linear Regression": pipeline_lr_II,
    "Random Forest": pipeline_rf_II,
    "XGBoost": pipeline_xgb_II
}

In [261]:
target_col = 'target'
feature_cols = numerical_cols + categorical_cols + binary_features

In [262]:
results = eval(df, pipelines_II, target_col, feature_cols)


=========== ITERATION 1 | TEST SEASON 2021 ===========

Training Linear Regression...


ValueError: A given column is not a column of the dataframe